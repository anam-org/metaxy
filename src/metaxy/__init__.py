from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
