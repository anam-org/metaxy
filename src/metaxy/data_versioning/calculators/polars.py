"""Polars implementation of data version calculator."""

from collections.abc import Callable
from typing import TYPE_CHECKING, Any

import narwhals as nw
import polars as pl
import polars_hash as plh

from metaxy.data_versioning.calculators.base import DataVersionCalculator
from metaxy.data_versioning.hash_algorithms import HashAlgorithm
from metaxy.utils.hashing import truncate_struct_column

if TYPE_CHECKING:
    from metaxy.models.feature_spec import BaseFeatureSpec, IDColumns
    from metaxy.models.plan import FeaturePlan


class PolarsDataVersionCalculator(DataVersionCalculator):
    """Calculates data versions using polars-hash.

    Accepts Narwhals LazyFrames and converts internally to Polars for hashing.
    Supports all hash functions available in polars-hash plugin.
    Default is xxHash64 for cross-database compatibility.
    """

    # Map HashAlgorithm enum to polars-hash functions
    _HASH_FUNCTION_MAP: dict[HashAlgorithm, Callable[[pl.Expr], pl.Expr]] = {
        HashAlgorithm.XXHASH64: lambda expr: expr.nchash.xxhash64(),  # pyright: ignore[reportAttributeAccessIssue]
        HashAlgorithm.XXHASH32: lambda expr: expr.nchash.xxhash32(),  # pyright: ignore[reportAttributeAccessIssue]
        HashAlgorithm.WYHASH: lambda expr: expr.nchash.wyhash(),  # pyright: ignore[reportAttributeAccessIssue]
        HashAlgorithm.SHA256: lambda expr: expr.chash.sha2_256(),  # pyright: ignore[reportAttributeAccessIssue]
        HashAlgorithm.MD5: lambda expr: expr.nchash.md5(),  # pyright: ignore[reportAttributeAccessIssue]
    }

    @property
    def supported_algorithms(self) -> list[HashAlgorithm]:
        """All algorithms supported by polars-hash."""
        return list(self._HASH_FUNCTION_MAP.keys())

    @property
    def default_algorithm(self) -> HashAlgorithm:
        """xxHash64 - fast and cross-database compatible."""
        return HashAlgorithm.XXHASH64

    def calculate_data_versions(
        self,
        joined_upstream: nw.LazyFrame[Any],
        feature_spec: "BaseFeatureSpec[IDColumns]",
        feature_plan: "FeaturePlan",
        upstream_column_mapping: dict[str, str],
        hash_algorithm: HashAlgorithm | None = None,
    ) -> nw.LazyFrame[Any]:
        """Calculate data_version using polars-hash.

        Args:
            joined_upstream: Narwhals LazyFrame with upstream data joined
            feature_spec: Feature specification
            feature_plan: Feature plan
            upstream_column_mapping: Maps upstream key -> column name
            hash_algorithm: Hash to use (default: xxHash64)

        Returns:
            Narwhals LazyFrame with data_version column added
        """
        algo = hash_algorithm or self.default_algorithm

        if algo not in self.supported_algorithms:
            raise ValueError(
                f"Hash algorithm {algo} not supported by PolarsDataVersionCalculator. "
                f"Supported: {self.supported_algorithms}"
            )

        # Convert Narwhals LazyFrame to Polars LazyFrame
        # Must collect first (LazyFrame doesn't have to_polars, only DataFrame does)
        pl_lazy = joined_upstream.collect().to_polars().lazy()

        hash_fn = self._HASH_FUNCTION_MAP[algo]

        # Build hash expressions for each field
        field_exprs = {}

        for field in feature_spec.fields:
            field_key_str = (
                field.key.to_string()
                if hasattr(field.key, "to_string")
                else "_".join(field.key)
            )

            field_deps = feature_plan.field_dependencies.get(field.key, {})

            # Build hash components
            components = [
                pl.lit(field_key_str),
                pl.lit(str(field.code_version)),
            ]

            # Add upstream data versions in deterministic order
            for upstream_feature_key in sorted(field_deps.keys()):
                upstream_fields = field_deps[upstream_feature_key]
                upstream_key_str = (
                    upstream_feature_key.to_string()
                    if hasattr(upstream_feature_key, "to_string")
                    else "_".join(upstream_feature_key)
                )

                data_version_col_name = upstream_column_mapping.get(
                    upstream_key_str, "data_version"
                )

                for upstream_field in sorted(upstream_fields):
                    upstream_field_str = (
                        upstream_field.to_string()
                        if hasattr(upstream_field, "to_string")
                        else "_".join(upstream_field)
                    )

                    components.append(
                        pl.lit(f"{upstream_key_str}/{upstream_field_str}")
                    )
                    components.append(
                        pl.col(data_version_col_name).struct.field(upstream_field_str)
                    )

            # Concatenate and hash
            concat_expr = plh.concat_str(*components, separator="|")
            hashed = hash_fn(concat_expr).cast(pl.Utf8)

            field_exprs[field_key_str] = hashed

        # Create data_version struct
        data_version_expr = pl.struct(**field_exprs)  # type: ignore[call-overload]

        result_pl = pl_lazy.with_columns(data_version_expr.alias("data_version"))

        # Convert back to Narwhals LazyFrame and apply truncation
        result_nw = nw.from_native(result_pl, eager_only=False)

        # Use truncate_struct_column to apply hash truncation if configured
        # This needs to be eager for the truncation function
        result_eager = result_nw.collect()
        result_truncated = truncate_struct_column(result_eager, "data_version")

        # Convert back to lazy
        return nw.from_native(result_truncated.to_native().lazy(), eager_only=False)
