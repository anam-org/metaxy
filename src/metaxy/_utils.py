import narwhals as nw
import polars as pl


def collect_to_polars(lazy_frame: nw.LazyFrame) -> pl.DataFrame:
    """Helper to collect a Narwhals LazyFrame and convert to Polars DataFrame.

    This handles all backend conversions (Polars, DuckDB/PyArrow, etc.) transparently.
    Used throughout tests for materializing query results.
    """
    return lazy_frame.collect().to_polars()
