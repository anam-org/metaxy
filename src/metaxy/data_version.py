"""Sample-level data version calculation using Polars expressions."""

import polars as pl
import polars_hash as plh


def calculate_sample_data_versions(
    upstream_data_versions: dict[str, pl.DataFrame],
    container_key: str,
    code_version: int,
    container_deps: dict[str, list[str]],
) -> pl.Expr:
    """Calculate sample-level data versions for a container.

    This function computes a Merkle tree hash for each sample by:
    1. Loading relevant upstream data versions for each sample
    2. Hashing them together (Merkle tree approach)
    3. Adding the current container's code version to the hash

    Args:
        upstream_data_versions: Dictionary mapping upstream feature keys to DataFrames.
            Each DataFrame should have a "data_version" column which is a pl.Struct
            with fields corresponding to containers for that feature.
        container_key: The key of the current container we're computing versions for.
        code_version: The code version of the current container.
        container_deps: Dictionary mapping upstream feature keys to lists of container
            keys that this container depends on.

    Returns:
        A Polars expression that computes the data version struct for each sample.
        The resulting struct will have a field for the current container.

    Example:
        >>> # Upstream feature "video" has containers ["frames", "audio"]
        >>> upstream_df = pl.DataFrame({
        ...     "sample_id": [1, 2, 3],
        ...     "data_version": [
        ...         {"frames": "abc123", "audio": "def456"},
        ...         {"frames": "abc124", "audio": "def457"},
        ...         {"frames": "abc125", "audio": "def458"},
        ...     ]
        ... })
        >>> upstream_data_versions = {"video": upstream_df}
        >>> container_deps = {"video": ["frames", "audio"]}
        >>>
        >>> # Calculate data version for our container
        >>> expr = calculate_sample_data_versions(
        ...     upstream_data_versions=upstream_data_versions,
        ...     container_key="processed",
        ...     code_version=1,
        ...     container_deps=container_deps,
        ... )
    """
    # Start with the container key and code version
    components = [
        pl.lit(container_key),
        pl.lit(str(code_version)),
    ]

    # Add upstream data versions in a deterministic order
    for feature_key in sorted(container_deps.keys()):
        container_keys = sorted(container_deps[feature_key])

        for upstream_container_key in container_keys:
            # Extract the specific container's data version from the upstream struct
            # Format: feature_key/container_key -> data_version
            components.append(pl.lit(f"{feature_key}/{upstream_container_key}"))
            components.append(
                pl.col("data_version").struct.field(upstream_container_key)
            )

    # Concatenate all components with a separator and hash
    # Use SHA256 via polars-hash
    data_version_expr = (
        plh.concat_str(*components, separator="|").chash.sha2_256().alias(container_key)
    )

    # Return as a struct with the container key as the field name
    return pl.struct(data_version_expr)


def calculate_feature_data_versions(
    upstream_data_versions: dict[str, pl.DataFrame],
    feature_containers: dict[str, int],
    feature_deps: dict[str, dict[str, list[str]]],
) -> pl.Expr:
    """Calculate sample-level data versions for all containers in a feature.

    Args:
        upstream_data_versions: Dictionary mapping upstream feature keys to DataFrames.
            Each DataFrame should have a "data_version" column which is a pl.Struct.
        feature_containers: Dictionary mapping container keys to their code versions
            for the current feature.
        feature_deps: Dictionary mapping container keys to their dependencies.
            Each dependency is a dict mapping upstream feature keys to lists of
            container keys.

    Returns:
        A Polars expression that computes a struct with all container data versions.

    Example:
        >>> upstream_data_versions = {
        ...     "video": pl.DataFrame({
        ...         "sample_id": [1, 2],
        ...         "data_version": [
        ...             {"frames": "abc", "audio": "def"},
        ...             {"frames": "ghi", "audio": "jkl"},
        ...         ]
        ...     })
        ... }
        >>> feature_containers = {
        ...     "processed": 1,
        ...     "augmented": 2,
        ... }
        >>> feature_deps = {
        ...     "processed": {"video": ["frames"]},
        ...     "augmented": {"video": ["frames", "audio"]},
        ... }
        >>> expr = calculate_feature_data_versions(
        ...     upstream_data_versions=upstream_data_versions,
        ...     feature_containers=feature_containers,
        ...     feature_deps=feature_deps,
        ... )
    """
    # Calculate data version for each container
    container_exprs = {}

    for container_key in sorted(feature_containers.keys()):
        code_version = feature_containers[container_key]
        container_deps = feature_deps.get(container_key, {})

        data_version_expr = calculate_sample_data_versions(
            upstream_data_versions=upstream_data_versions,
            container_key=container_key,
            code_version=code_version,
            container_deps=container_deps,
        )

        # Extract the field from the struct
        container_exprs[container_key] = data_version_expr.struct.field(container_key)

    # Combine all container data versions into a single struct
    return pl.struct(**container_exprs)


def merge_upstream_data_versions(
    sample_df: pl.DataFrame,
    upstream_data_versions: dict[str, pl.DataFrame],
    feature_key_col: str = "sample_id",
) -> pl.DataFrame:
    """Merge upstream data versions into the sample DataFrame.

    This helper function joins all upstream data version DataFrames with the
    current sample DataFrame, creating columns like "video__data_version",
    "audio__data_version", etc.

    Args:
        sample_df: The DataFrame containing samples for the current feature.
        upstream_data_versions: Dictionary mapping upstream feature keys to their
            DataFrames with data versions.
        feature_key_col: The column name to use for joining (default: "sample_id").

    Returns:
        DataFrame with all upstream data versions joined.
    """
    result_df = sample_df

    for feature_key, upstream_df in upstream_data_versions.items():
        # Rename the data_version column to avoid conflicts
        upstream_df_renamed = upstream_df.select(
            [
                pl.col(feature_key_col),
                pl.col("data_version").alias(f"{feature_key}__data_version"),
            ]
        )

        result_df = result_df.join(
            upstream_df_renamed,
            on=feature_key_col,
            how="left",
        )

    return result_df
