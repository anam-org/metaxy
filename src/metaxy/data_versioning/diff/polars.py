"""Polars implementation of metadata diff resolver."""

import polars as pl

from metaxy.data_versioning.diff.base import DiffResult, MetadataDiffResolver


class PolarsDiffResolver(MetadataDiffResolver):
    """Identifies changed rows using Polars operations.

    Strategy:
    - Categorizes changes into added, changed, and removed
    - Uses LEFT/RIGHT JOINs to identify each category
    - Materializes once and splits into three DataFrames (efficient)
    """

    def find_changes(
        self,
        target_versions: pl.LazyFrame,
        current_metadata: pl.LazyFrame | None,
    ) -> DiffResult:
        """Find all changes between target and current.

        Args:
            target_versions: LazyFrame with calculated data_versions
            current_metadata: LazyFrame with current metadata, or None

        Returns:
            DiffResult with three materialized DataFrames
        """
        # Select only sample_id and data_version from target_versions
        # (it may have intermediate joined columns from upstream)
        target_versions = target_versions.select(["sample_id", "data_version"])

        if current_metadata is None:
            # No existing metadata - all target rows are new, nothing removed/changed
            empty_df = pl.DataFrame({"sample_id": [], "data_version": []})
            return DiffResult(
                added=target_versions.collect(),
                changed=empty_df,
                removed=empty_df,
            )

        # Join target with current to categorize changes
        # Keep only sample_id and data_version from current for comparison
        current_comparison = current_metadata.select(
            ["sample_id", pl.col("data_version").alias("__current_data_version")]
        )

        # LEFT JOIN target with current
        compared = target_versions.join(
            current_comparison,
            on="sample_id",
            how="left",
        )

        # Build lazy queries for each category
        added_lazy = (
            compared.filter(pl.col("__current_data_version").is_null())
            .drop("__current_data_version")
            .select(["sample_id", "data_version"])
        )

        changed_lazy = (
            compared.filter(
                pl.col("__current_data_version").is_not_null()
                & (pl.col("data_version") != pl.col("__current_data_version"))
            )
            .drop("__current_data_version")
            .select(["sample_id", "data_version"])
        )

        removed_lazy = current_metadata.join(
            target_versions.select("sample_id"),
            on="sample_id",
            how="anti",
        ).select(["sample_id", "data_version"])

        # Collect all three at once (single execution for shared computations)
        added_df, changed_df, removed_df = pl.collect_all(
            [added_lazy, changed_lazy, removed_lazy]
        )

        return DiffResult(
            added=added_df,
            changed=changed_df,
            removed=removed_df,
        )
