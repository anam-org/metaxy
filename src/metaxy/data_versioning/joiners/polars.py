"""Polars implementation of upstream joiner."""

from typing import TYPE_CHECKING

import polars as pl

from metaxy.data_versioning.joiners.base import UpstreamJoiner

if TYPE_CHECKING:
    from metaxy.models.feature_spec import FeatureSpec
    from metaxy.models.plan import FeaturePlan


class PolarsJoiner(UpstreamJoiner[pl.LazyFrame]):
    """Joins upstream features using Polars LazyFrames.

    Type Parameters:
        TRef = pl.LazyFrame

    Strategy:
    - Starts with first upstream feature
    - Sequentially inner joins remaining upstream features on sample_id
    - Renames data_version columns to avoid conflicts
    - All operations are lazy (no materialization until collect)
    """

    def join_upstream(
        self,
        upstream_refs: dict[str, pl.LazyFrame],
        feature_spec: "FeatureSpec",
        feature_plan: "FeaturePlan",
    ) -> tuple[pl.LazyFrame, dict[str, str]]:
        """Join upstream LazyFrames together.

        Args:
            upstream_refs: Dict of upstream feature key -> LazyFrame
            feature_spec: Feature specification
            feature_plan: Feature plan

        Returns:
            (joined LazyFrame, column mapping)
        """
        if not upstream_refs:
            # No upstream dependencies - source feature
            # Return empty LazyFrame with just sample_id column
            # The calculator will generate data_versions based only on feature_version
            return pl.LazyFrame({"sample_id": []}), {}

        # Start with the first upstream feature
        upstream_keys = sorted(upstream_refs.keys())
        first_key = upstream_keys[0]

        # Start with first upstream, rename its data_version column
        col_name_first = f"__upstream_{first_key}__data_version"
        joined = upstream_refs[first_key].select(
            ["sample_id", pl.col("data_version").alias(col_name_first)]
        )

        upstream_mapping = {first_key: col_name_first}

        # Join remaining upstream features
        for upstream_key in upstream_keys[1:]:
            col_name = f"__upstream_{upstream_key}__data_version"

            upstream_renamed = upstream_refs[upstream_key].select(
                ["sample_id", pl.col("data_version").alias(col_name)]
            )

            joined = joined.join(
                upstream_renamed,
                on="sample_id",
                how="inner",  # Only sample_ids present in ALL upstream
            )

            upstream_mapping[upstream_key] = col_name

        return joined, upstream_mapping
