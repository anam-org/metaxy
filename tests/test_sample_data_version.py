"""Tests for sample-level data version calculation."""

import polars as pl
from syrupy.assertion import SnapshotAssertion

from metaxy.data_version import (
    calculate_feature_data_versions,
    calculate_sample_data_versions,
)


def test_calculate_sample_data_versions_single_upstream(
    snapshot: SnapshotAssertion,
) -> None:
    """Test calculating data versions with a single upstream feature."""
    # Create upstream data with two containers
    upstream_df = pl.DataFrame(
        {
            "sample_id": [1, 2, 3],
            "data_version": [
                {"frames": "abc123", "audio": "def456"},
                {"frames": "abc124", "audio": "def457"},
                {"frames": "abc125", "audio": "def458"},
            ],
        }
    )

    upstream_data_versions = {"video": upstream_df}
    container_deps = {"video": ["frames", "audio"]}

    # Calculate data version expression
    expr = calculate_sample_data_versions(
        upstream_data_versions=upstream_data_versions,
        container_key="processed",
        code_version=1,
        container_deps=container_deps,
    )

    # Apply to the upstream DataFrame
    result = upstream_df.select(
        [
            pl.col("sample_id"),
            expr.alias("computed_version"),
        ]
    )

    # Check that we got a struct
    assert result.schema["computed_version"] == pl.Struct(
        [pl.Field("processed", pl.String)]
    )

    # Check that all samples have different data versions
    versions = result["computed_version"].struct.field("processed")
    assert len(versions.unique()) == 3

    # Check that versions are non-empty strings
    assert all(len(v) > 0 for v in versions)

    # Snapshot the actual hash values to ensure stability
    assert versions.to_list() == snapshot


def test_calculate_sample_data_versions_multiple_upstream(
    snapshot: SnapshotAssertion,
) -> None:
    """Test calculating data versions with multiple upstream features."""
    # Create two upstream features
    video_df = pl.DataFrame(
        {
            "sample_id": [1, 2, 3],
            "data_version": [
                {"frames": "video1", "audio": "audio1"},
                {"frames": "video2", "audio": "audio2"},
                {"frames": "video3", "audio": "audio3"},
            ],
        }
    )

    metadata_df = pl.DataFrame(
        {
            "sample_id": [1, 2, 3],
            "data_version": [
                {"tags": "tags1"},
                {"tags": "tags2"},
                {"tags": "tags3"},
            ],
        }
    )

    # Join upstream data versions
    video_df.join(
        metadata_df.select(
            [
                pl.col("sample_id"),
                pl.col("data_version").alias("metadata_data_version"),
            ]
        ),
        on="sample_id",
    )

    # For this test, we need to handle multiple upstream features differently
    # Let's test with just video for now
    upstream_data_versions = {"video": video_df}
    container_deps = {"video": ["frames"]}

    expr = calculate_sample_data_versions(
        upstream_data_versions=upstream_data_versions,
        container_key="processed",
        code_version=1,
        container_deps=container_deps,
    )

    result = video_df.select(
        [
            pl.col("sample_id"),
            expr.alias("computed_version"),
        ]
    )

    versions = result["computed_version"].struct.field("processed")
    assert len(versions.unique()) == 3

    # Snapshot the hash values
    assert versions.to_list() == snapshot


def test_calculate_sample_data_versions_deterministic(
    snapshot: SnapshotAssertion,
) -> None:
    """Test that data version calculation is deterministic."""
    upstream_df = pl.DataFrame(
        {
            "sample_id": [1, 1],  # Same sample ID
            "data_version": [
                {"frames": "same_hash", "audio": "same_hash"},
                {"frames": "same_hash", "audio": "same_hash"},
            ],
        }
    )

    upstream_data_versions = {"video": upstream_df}
    container_deps = {"video": ["frames", "audio"]}

    expr = calculate_sample_data_versions(
        upstream_data_versions=upstream_data_versions,
        container_key="processed",
        code_version=1,
        container_deps=container_deps,
    )

    result = upstream_df.select(
        [
            expr.alias("computed_version"),
        ]
    )

    versions = result["computed_version"].struct.field("processed")
    # Both should have the same hash since inputs are identical
    assert versions[0] == versions[1]

    # Snapshot the deterministic hash value
    assert versions[0] == snapshot


def test_calculate_sample_data_versions_code_version_change(
    snapshot: SnapshotAssertion,
) -> None:
    """Test that changing code version produces different data versions."""
    upstream_df = pl.DataFrame(
        {
            "sample_id": [1],
            "data_version": [
                {"frames": "abc123"},
            ],
        }
    )

    upstream_data_versions = {"video": upstream_df}
    container_deps = {"video": ["frames"]}

    # Calculate with code version 1
    expr_v1 = calculate_sample_data_versions(
        upstream_data_versions=upstream_data_versions,
        container_key="processed",
        code_version=1,
        container_deps=container_deps,
    )

    result_v1 = upstream_df.select(
        [
            expr_v1.alias("computed_version"),
        ]
    )

    # Calculate with code version 2
    expr_v2 = calculate_sample_data_versions(
        upstream_data_versions=upstream_data_versions,
        container_key="processed",
        code_version=2,
        container_deps=container_deps,
    )

    result_v2 = upstream_df.select(
        [
            expr_v2.alias("computed_version"),
        ]
    )

    version_v1 = result_v1["computed_version"].struct.field("processed")[0]
    version_v2 = result_v2["computed_version"].struct.field("processed")[0]

    # Different code versions should produce different data versions
    assert version_v1 != version_v2

    # Snapshot both versions to ensure they remain stable
    assert {"v1": version_v1, "v2": version_v2} == snapshot


def test_calculate_feature_data_versions(snapshot: SnapshotAssertion) -> None:
    """Test calculating data versions for all containers in a feature."""
    upstream_df = pl.DataFrame(
        {
            "sample_id": [1, 2],
            "data_version": [
                {"frames": "abc", "audio": "def"},
                {"frames": "ghi", "audio": "jkl"},
            ],
        }
    )

    upstream_data_versions = {"video": upstream_df}

    feature_containers = {
        "processed": 1,
        "augmented": 2,
    }

    feature_deps = {
        "processed": {"video": ["frames"]},
        "augmented": {"video": ["frames", "audio"]},
    }

    expr = calculate_feature_data_versions(
        upstream_data_versions=upstream_data_versions,
        feature_containers=feature_containers,
        feature_deps=feature_deps,
    )

    result = upstream_df.select(
        [
            pl.col("sample_id"),
            expr.alias("data_version"),
        ]
    )

    # Check structure
    assert "data_version" in result.columns
    data_version_type = result.schema["data_version"]
    assert isinstance(data_version_type, pl.Struct)

    # Check that both containers are present
    fields = {field.name for field in data_version_type.fields}
    assert fields == {"processed", "augmented"}

    # Check that versions are different for different samples
    processed_versions = result["data_version"].struct.field("processed")
    assert len(processed_versions.unique()) == 2

    augmented_versions = result["data_version"].struct.field("augmented")
    assert len(augmented_versions.unique()) == 2

    # Check that processed and augmented have different versions
    # (since augmented depends on more containers)
    assert processed_versions[0] != augmented_versions[0]

    # Snapshot all hash values (convert to dict for readability)
    hash_dict = {
        "processed": processed_versions.to_list(),
        "augmented": augmented_versions.to_list(),
    }
    assert hash_dict == snapshot
