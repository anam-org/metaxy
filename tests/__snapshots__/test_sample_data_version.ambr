# serializer version: 1
# name: test_calculate_feature_data_versions
  dict({
    'augmented': list([
      '7c720a2d2a26d1ba28f540da1ce67215c58d1bce016b39485ca3d1648dbb3f50',
      'ea8d2d10cea126876bae102d689aa3ec313c982c7bacea4ad5ab02d56777d78d',
    ]),
    'processed': list([
      '244dbc8db0e556e38bd7bd234219e7e5da68dd318a37ef8e6134c9ea31f09238',
      'd709c55fa1b4ac42c97e556e1123cc4a85ed7f2eb93a3ae6d9b8602dddd859fb',
    ]),
  })
# ---
# name: test_calculate_sample_data_versions_code_version_change
  dict({
    'v1': 'f35ce48d9cfba8a23415863b518f6f81d0cf24553dc3540f4a62f05783ef4217',
    'v2': '11966375903503534768c8f67099bd404e89810944a974ea289ff7d2bef35f04',
  })
# ---
# name: test_calculate_sample_data_versions_deterministic
  'd02a0e24362904e5bee36e20c3cd6a639840b7e339c34dca1eaea7f644fa4cc2'
# ---
# name: test_calculate_sample_data_versions_multiple_upstream
  list([
    '26852f56de452ba1cd7224e7bfce22805c5eed54f5691ee77fcc4cfeb18d4970',
    'ea0ac32cf60ce13cfee50c099c3950d906ed55f349f1186ab7aa217dd3aed3dc',
    'b29bb555048efb739b1b2f535dae297f76a87fe4cb944f607131e3b18e262c50',
  ])
# ---
# name: test_calculate_sample_data_versions_single_upstream
  list([
    'fc0ba303d12aecaa868ac0b36cfed5d7635ff185871a0b96d8b338d222e8d3aa',
    '0a33c360b5332c4b95ecfae8a5ba8899e9d75cf2111bf6e126d0ec90c67ba580',
    '9addc04b84dcfbe4c040da967d37fd6980c6984bf629d5c1008ec728f21acc17',
  ])
# ---
